momepy/preprocessing.py

```diff
@@ -740,3 +748,240 @@ def _get_extrapolated_line(coords, tolerance, point=False):
+    """
+    # calculate parameters
+    gdf["area"] = gdf.geometry.area
+    gdf["circom"] = CircularCompactness(gdf, "area").series
+    
+    #selecting round about polygons based on compactness
+    rab = gdf[gdf.circom > circom_threshold]
+    #exclude those above the area threshold
+    area_threshold = gdf.area.quantile(perc_area_threshold)
+    rab = rab[rab.area < area_threshold]
+    
+    if include_adjacent == True :
+        #selecting the adjacent areas that are of smaller than itself
+        rab_adj = gpd.sjoin(gdf, rab, predicate = 'intersects')
+        rab_adj = rab_adj[rab_adj.area_right >= rab_adj.area_left]
+        
+        adj_long_polys = []
+        #iterate through original rabs
+        for row in rab.itertuples() :
+            #creating a distance (diameter) for limiting the max dist of adjacent
+            minx, _ , maxx, _ = row.geometry.bounds 
+            rab_diameter = maxx - minx
+            
+            #creating the subgroup that could be affected
+            rab_group = rab_adj[rab_adj.index_right==row[0]]
+            for rrow in rab_group.itertuples():
+                hdist = row.geometry.hausdorff_distance(rrow.geometry)
+
+                if hdist > rab_diameter :
+                    adj_long_polys.append(rrow[0]) #if hausdorff_distance is larger than the diameter add to list
+           
+        rab_plus = rab_adj.drop(adj_long_polys)
+        
+    else:
+        rab['index_right'] = rab.index
+        rab_plus = rab
+    
+    rab_plus = rab_plus[['geometry', 'index_right']] #only keeping relevant fields
+    
+    return rab_plus
+
+def _rabs_center_points(gdf, center_type = 'centroid'):
+    """
+    From a selection of round abouts, returns an aggregated GeoDataFrame 
+    per round about with extra column with center_type. 
+    """
+    #creating a multipolygon per RAB (as opposed to dissolving) of the entire composition of the RAB
+    # temporary DataFrame where geometries is the array of pygeos geometries
+    # for aggregation to work properly
+    tmp = pd.DataFrame(gdf.copy())
+    tmp['geometry'] = tmp.geometry.values.data
+
+    rab_plus = gpd.GeoDataFrame(
+            tmp.groupby('index_right').geometry.apply(pygeos.multipolygons).rename('geometry'), crs= gdf.crs)
+    # verifying that all geometries are valid
+    rab_plus['geometry'] = rab_plus.apply(lambda row : make_valid(row.geometry) if not row.geometry.is_valid else row.geometry, axis = 1)
+    
+    if center_type == 'centroid' :
+        #geometry centroid of the actual circle
+        rab_plus['center_pt'] = gdf[gdf.index == gdf.index_right].geometry.centroid
+    
+    elif center_type == 'mean':
+        # mean geometry
+        ls_xy = [g.exterior.coords.xy for g in rab_plus.geometry] #extracting the points
+        
+        ls_xy = [point for polygon in multipolygon for point in polygon.exterior.coords[:-1]]
+        
+        mean_pts = [geometry.Point(np.mean(xy[0]),np.mean(xy[1])) for xy in ls_xy]
```

Comment:
get arrays and create points with `points_from_xy` or pygeos. always try to use a vectorised solution if there is one.